`valid-typeof` with `requireStringLiterals` should allow two `typeof` comparisons

**What version of ESLint are you using?**
v3.4.0

**What parser (default, Babel-ESLint, etc.) are you using?**
default

**Please show your full configuration:**
just `'valid-typeof': ['error', { requireStringLiterals: true }]`

**What did you do? Please include the actual source code causing the issue.**
`echo 'typeof foo === typeof bar' | eslint --stdin --no-eslintrc --rule valid-typeof:'[2,{requireStringLiterals: true}]'`

**What did you expect to happen?**
No error.

**What actually happened? Please include the actual, raw output from ESLint.**
```sh
text>
  1:1   error  Typeof comparisons should be to string literals  valid-typeof
  1:16  error  Typeof comparisons should be to string literals  valid-typeof

✖ 2 problems (2 errors, 0 warnings)
```

--- a/docs/rules/valid-typeof.md
+++ b/docs/rules/valid-typeof.md
@@ -10,7 +10,7 @@ This rule enforces comparing `typeof` expressions to valid string literals.
 
 This rule has an object option:
 
-* `"requireStringLiterals": true` requires `typeof` expressions to only be compared to string literals, and disallows comparisons to any other value.
+* `"requireStringLiterals": true` requires `typeof` expressions to only be compared to string literals or other `typeof` expressions, and disallows comparisons to any other value.
 
 Examples of **incorrect** code for this rule:
 
@@ -43,7 +43,6 @@ typeof baz === "strnig"
 typeof qux === "some invalid type"
 typeof baz === anotherVariable
 typeof foo == 5
-typeof bar === typeof qux
 ```
 
 Examples of **correct** code with the `{ "requireStringLiterals": true }` option:
@@ -52,6 +51,7 @@ Examples of **correct** code with the `{ "requireStringLiterals": true }` option
 typeof foo === "undefined"
 typeof bar == "object"
 typeof baz === "string"
+typeof bar === typeof qux
 ```
 
 ## When Not To Use It

--- a/lib/rules/valid-typeof.js
+++ b/lib/rules/valid-typeof.js
@@ -34,6 +34,17 @@ module.exports = {
         const VALID_TYPES = ["symbol", "undefined", "object", "boolean", "number", "string", "function"],
             OPERATORS = ["==", "===", "!=", "!=="];
 
+        const requireStringLiterals = context.options[0] && context.options[0].requireStringLiterals;
+
+        /**
+        * Determines whether a node is a typeof expression.
+        * @param {ASTNode} node The node
+        * @returns {boolean} `true` if the node is a typeof expression
+        */
+        function isTypeofExpression(node) {
+            return node.type === "UnaryExpression" && node.operator === "typeof";
+        }
+
         //--------------------------------------------------------------------------
         // Public
         //--------------------------------------------------------------------------
@@ -41,7 +52,7 @@ module.exports = {
         return {
 
             UnaryExpression(node) {
-                if (node.operator === "typeof") {
+                if (isTypeofExpression(node)) {
                     const parent = context.getAncestors().pop();
 
                     if (parent.type === "BinaryExpression" && OPERATORS.indexOf(parent.operator) !== -1) {
@@ -51,7 +62,7 @@ module.exports = {
                             if (VALID_TYPES.indexOf(sibling.value) === -1) {
                                 context.report(sibling, "Invalid typeof comparison value.");
                             }
-                        } else if (context.options[0] && context.options[0].requireStringLiterals) {
+                        } else if (requireStringLiterals && !isTypeofExpression(sibling)) {
                             context.report(sibling, "Typeof comparisons should be to string literals.");
                         }
                     }

--- a/tests/lib/rules/valid-typeof.js
+++ b/tests/lib/rules/valid-typeof.js
@@ -55,6 +55,10 @@ ruleTester.run("valid-typeof", rule, {
         {
             code: "var baz = typeof foo + 'thing'",
             options: [{ requireStringLiterals: true }]
+        },
+        {
+            code: "typeof foo === typeof bar",
+            options: [{ requireStringLiterals: true }]
         }
     ],
 

